update: add about defining Kotlin compiler execution strategy

PetrakovichVictoria opened the pull request.

ALikhachev commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+|----------------|-----------------------------------|-------------------------------------|----------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *Enabled by default*. Can be shared between different Gradle daemons |
+| In process     | Inside Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
+| Out of process | In separate forked process        | No                                  | -                                                                    |
```
```suggestion
| Out of process | In a separate forked process        | No                                  | -                                                                    |
```

ALikhachev commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                |
+|----------------|-----------------------------------|-------------------------------------|----------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *Enabled by default*. Can be shared between different Gradle daemons |
+| In process     | Inside Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
```
```suggestion
| In process     | Inside the Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
```

ALikhachev commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ is a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```
```suggestion
_Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```

Tapchicoma commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+The available values for the `compilerExecutionStrategy` task property are:
+1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`
+2. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.OUT_OF_PROCESS`
+3. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
```
Probably swap it with 1. position

PetrakovichVictoria commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+3. `daemon` (default)
+
+The available values for the `compilerExecutionStrategy` task property are:
+1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`
```
```suggestion
1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
```

PetrakovichVictoria commented:
@koshachy @p7nov I need a tech writers' review.
I fixed all the remarks from the tech review.
The text is pretty much the same as in the "What's new", just added the definition of the Kotlin compiler strategy.

p7nov commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy
+
+// Some code
```
No "some code" please. Either check how we show Gradle scripts with imports in other examples on this page or consider using FQNs in the code.

p7nov reviewed (changes requested):
The part after the table is written in the What's new style, which isn't suitable for docs. Please rewrite it so that it isn't tied to the specific release. You can mention old things and versions to preserve compatibility, but do it in some isolated parts so that you can remove them instantly when these things are deprecated.
The doc should be "timeless" and relevant no matter when people read it.

PetrakovichVictoria commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```
Agree about vague.

PetrakovichVictoria commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+
+// ...
+
+tasks.withType(KotlinCompile.class)
```
```suggestion
tasks.withType(KotlinCompile)
```
```suggestion
tasks.withType(KotlinCompile.class)
```

p7nov commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ describes where Kotlin compiler is executed and if incremental compilation is supported in each case.
```
Maybe `defines`? Though the current wording seems OK too.

PetrakovichVictoria replied:
Defines is better indeed, thank you.

p7nov commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+|----------------|-----------------------------------|-------------------------------------|------------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *The default strategy*. Can be shared between different Gradle daemons |
+| In process     | Inside the Gradle daemon process  | No                                  | May share heap with the Gradle daemon                                  |
+| Out of process | In a separate forked process      | No                                  | -                                                                      |
```
Not sure if `forked` adds meaning here. Does it have any practical use?

p7nov replied:
For me, it only raised questions "Forked from what? What should I do with the fact that it was forked?"

PetrakovichVictoria replied:
@ALikhachev could you help, please?

ALikhachev replied:
Yep, I guess we can remove `forked`. It's a technical detail
I'd suggest to rephrase it to be more clear
like
```suggestion
| Out of process | In a separate process for each call      | No                                  | -                                                                      |
```
So the new description emphasises that a new process will be created for each compilation task.

p7nov commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+To define a Kotlin compiler execution strategy, you can use one of the following properties:
+* The `kotlin.compiler.execution.strategy` Gradle property.
+* The `compilerExecutionStrategy` compile task property.
+* The `-Dkotlin.compiler.execution.strategy` system property, which will be removed in future releases. 
```
Let's put the last thought in a separate note. Also, is it deprecated in 1.6.20? If yes, say this explicitly.

PetrakovichVictoria replied:
>is it deprecated in 1.6.20?

I don't have such info, @ALikhachev could you help, please?

ALikhachev replied:
It's not deprecated yet, and we don't show any warnings on it's usages. We will add deprecation warning in 1.7 and then remove it in 1.8.

p7nov reviewed (approved):
Fine by me now, a couple of minor comments.

PetrakovichVictoria commented on docs/topics/gradle.md:
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ describes where Kotlin compiler is executed and if incremental compilation is supported in each case.
```
```suggestion
_Kotlin compiler execution strategy_ defines where Kotlin compiler is executed and if incremental compilation is supported in each case.
```
